src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,108 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double SHOULD_GROW = 1.5;
+    private static final int MIN_INDEX_VALUE = 0;
+    private static final int DIFFERENCE_IN_SIZES = 1;
+    private int size;
+    private Object[] objects;
+
+    public ArrayList() {
+        objects = new Object[DEFAULT_CAPACITY];
+    }
 
     @Override
     public void add(T value) {
-
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < objects.length) {
+            if (index <= size && index >= MIN_INDEX_VALUE) {
+                System.arraycopy(objects, index, objects,
+                        index + DIFFERENCE_IN_SIZES, size - index);
+                objects[index] = value;
+                size++;
+            } else {
+                throw new ArrayListIndexOutOfBoundsException("Index is out of ArrayList bounds");
+            }
+        } else {
+            objects = grow();
+            add(value,index);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list.size() + (size - DIFFERENCE_IN_SIZES) < objects.length) {
+            for (int i = MIN_INDEX_VALUE; i < list.size(); i++) {
+                add(list.get(i));
+            }
+        } else {
+            objects = grow();
+            addAll(list);
+        }
+    }
 
+    private Object[] grow() {
+        Object[] biggerOne = new Object[(int)(objects.length * SHOULD_GROW)];
+        System.arraycopy(objects, MIN_INDEX_VALUE, biggerOne, MIN_INDEX_VALUE, size);
+        return biggerOne;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < size && index >= MIN_INDEX_VALUE) {
+            return (T) objects[index];
+        } else {
```

Comment:
all of the following methods we can implement without `else`. try to not complicate your code